ast/asdl_rs.py

```diff
@@ -243,84 +253,111 @@ def visitProduct(self, product, name, depth):
-            return "<U = ()>", "<U>"
-        else:
-            return "", ""
 
-class MapAstImplVisitor(EmitVisitor):
-    def __init__(self, file, typeinfo):
-        self.typeinfo = typeinfo
-        super().__init__(file)
+class FoldTraitDefVisitor(TypeInfoEmitVisitor):
+    def visitModule(self, mod, depth):
+        self.emit("pub trait Fold<U> {", depth)
+        self.emit("type TargetU;", depth + 1)
+        self.emit("type Error;", depth + 1)
+        self.emit("fn map_user(&mut self, user: U) -> Result<Self::TargetU, Self::Error>;", depth + 2)
+        for dfn in mod.dfns:
+            self.visit(dfn, depth + 2)
+        self.emit("}", depth)
 
-    def get_generics(self, typ):
-        if self.typeinfo[typ].has_userdata:
-            return "<T>", "<U>"
-        else:
-            return "", ""
+    def visitType(self, type, depth):
+        name = type.name
+        apply_u, apply_target_u = self.get_generics(name, "U", "Self::TargetU")
+        enumname = get_rust_type(name)
+        self.emit(f"fn fold_{name}(&mut self, node: {enumname}{apply_u}) -> Result<{enumname}{apply_target_u}, Self::Error> {{", depth)
+        self.emit(f"fold_{name}(self, node)", depth + 1)
+        self.emit("}", depth)
 
-    def visitModule(self, mod):
+
+class FoldImplVisitor(TypeInfoEmitVisitor):
+    def visitModule(self, mod, depth):
+        self.emit("fn fold_located<U, F: Fold<U> + ?Sized, T, MT>(folder: &mut F, node: Located<T, U>, f: impl FnOnce(&mut F, T) -> Result<MT, F::Error>) -> Result<Located<MT, F::TargetU>, F::Error> {", depth)
+        self.emit("Ok(Located { custom: folder.map_user(node.custom)?, location: node.location, node: f(folder, node.node)? })", depth + 1)
+        self.emit("}", depth)
         for dfn in mod.dfns:
-            self.visit(dfn)
+            self.visit(dfn, depth)
 
     def visitType(self, type, depth=0):
         self.visit(type.value, type.name, depth)
 
     def visitSum(self, sum, name, depth):
-        apply_t, apply_u = self.get_generics(name)
+        apply_t, apply_u, apply_target_u = self.get_generics(name, "T", "U", "F::TargetU")
         enumname = get_rust_type(name)
-        if sum.attributes:
-            enumname += "Kind"
+        is_located = bool(sum.attributes)
 
-        if not apply_t:
-            self.emit(f"no_user!({enumname});", depth)
-            return
-
-
-        self.emit(f"impl<T, U> MapAst<T, U> for {enumname}{apply_t} {{", depth)
+        self.emit(f"impl<T, U> Foldable<T, U> for {enumname}{apply_t} {{", depth)
         self.emit(f"type Mapped = {enumname}{apply_u};", depth + 1)
-        self.emit("fn try_map_ast<E, F: FnMut(T) -> Result<U, E>>(self, f: &mut F) -> Result<Self::Mapped, E> {", depth + 1)
-        self.emit("match self {", depth + 2)
+        self.emit("fn fold<F: Fold<T, TargetU = U> + ?Sized>(self, folder: &mut F) -> Result<Self::Mapped, F::Error> {", depth + 1)
+        self.emit(f"fold_{name}(folder, self)", depth + 2)
```

Comment:
Shouldn't we use `folder.fold_{name}(self)` here? Otherwise, overriding the default implementation would not be useful as the default one would be called.

Also, I think we should add test cases for nested structure, for the constant propagation test. As the bug mentioned above is not revealed by current test case.

Reply:
Oh, very good point, thanks.